plugins/inputs/statsd/datadog.go

```
@@ -0,0 +1,173 @@
+	m.fields = make(map[string]interface{}, 9)
+	m.fields["alert-type"] = "info" // default event type
+	m.fields["text"] = uncommenter.Replace(string(rawText))
+	m.tags["source"] = defaultHostname
+	m.fields["priority"] = priorityNormal
+	m.ts = now
+	if len(message) == 0 {
+		s.events = append(s.events, m)
+		return nil
+	}
+
+	if len(message) > 1 {
+		rawMetadataFields := strings.Split(message[1:], "|")
+		for i := range rawMetadataFields {
+			if len(rawMetadataFields[i]) < 2 {
+				log.Printf("W! [inputs.statsd] too short metadata field")
+			}
+			switch rawMetadataFields[i][:2] {
+			case "d:":
+				ts, err := strconv.ParseInt(rawMetadataFields[i][2:], 10, 64)
+				if err != nil {
+					log.Printf("W! [inputs.statsd] skipping timestamp: %s", err)
+					continue
+				}
+				m.fields["ts"] = ts
+			case "p:":
+				switch rawMetadataFields[i][2:] {
+				case priorityLow:
+					m.fields["priority"] = priorityLow
+				case priorityNormal: // we already used this as a default
+				default:
+					log.Printf("W! [inputs.statsd] skipping priority")
+					continue
+				}
+			case "h:":
+				m.tags["source"] = rawMetadataFields[i][2:]
+			case "t:":
+				switch rawMetadataFields[i][2:] {
+				case "error":
+					m.fields["alert-type"] = "error"
+				case "warning":
+					m.fields["alert-type"] = "warning"
+				case "success":
+					m.fields["alert-type"] = "success"
+				case "info": // already set for info
+				default:
+					log.Printf("W! [inputs.statsd] skipping alert type")
+					continue
+				}
+			case "k:":
+				// TODO(docmerlin): does this make sense?
+				m.tags["aggregation-key"] = rawMetadataFields[i][2:]
+			case "s:":
+				m.fields["source-type-name"] = rawMetadataFields[i][2:]
+			default:
+				if rawMetadataFields[i][0] == '#' {
+					parseDataDogTags(m.tags, rawMetadataFields[i][1:])
+				} else {
+					log.Printf("W! [inputs.statsd] unknown metadata type: '%s'", rawMetadataFields[i])
```

@danielnelson so, the original source code would optimistically continue parsing and warn about bad values within the even message. 

Does this fit telegraf?  Should these warnings be errors or are logs ok?

---

I think if the message is malformed an error should be returned up from here and then logged.  Doing a best effort parse doesn't make sense unless we find that the datadog agent regularly emits bad messages.

---

Telegraf generally tries to ignore errors and keep working where it can.